superset/dashboards/api.py

```diff
@@ -342,6 +342,11 @@ def get(
               $ref: '#/components/responses/404'
         """
         result = self.dashboard_get_response_schema.dump(dash)
+        if security_manager.is_guest_user():
```

Comment:
@jfrag1 we should adhere to the "shift left" mentality and filter out inaccessible fields at the DAO layer as opposed to the API. This ensures we adhere to the DRY principle avoiding the need to repeat (and/or miss) said logic.

Reply:
@john-bodley Thanks for reviewing, but I'm not sure how we're supposed to properly filter out certain fields of a record in the DAO layer.  As far as I'm aware the security logic in the DAO layer is only "row-level", protecting entire objects, there's no column-level protections.  Are you suggesting in the appropriate DAO method do something like
```python
dashboard = db.session.query(...
if is_guest:
    dashboard.owners = []
    dashboard.changed_by = None
```
I don't like this since it introduces the possibility of these changes accidentally being committed, but please let me know if you had something else in mind or if this pattern does exist currently but I'm just not aware of it.

Reply:
@betodealmeida @dpgaspar @michael-s-molina @villebro what are your thoughts on where sanitization of the data response should occur? Note this doesn't seem to be API specific and thus I'm not sure we should explicitly be defining said logic there.

Reply:
Implementing this at the API layer is problematic because there's no single point of verification and also because you will query data that you don't have access to and filter it out later, wasting resources (imagine a table with many columns and you only have access to one column). This will also be true for the suggested code below:

```
dashboard = db.session.query(...
if is_guest:
    dashboard.owners = []
    dashboard.changed_by = None
```

Considering our current architecture, where we offer a [RLS (Row Level Security)](https://docs.preset.io/docs/row-level-security-rls) solution instead of depending on the database layer for these types of controls, the correct way of implementing this is to augment our security model to also support [Column Level Security (CLS)](https://blog.fabric.microsoft.com/en-us/blog/announcing-column-level-row-level-security-for-fabric-warehouse-sql-endpoint/) or even [Object Level Security (OLS)](https://learn.microsoft.com/en-us/power-bi/enterprise/service-admin-ols).

> I don't like this since it introduces the possibility of these changes accidentally being committed

Ideally, you won't query for what you don't have access to, eliminating this concern because the values won't be in the session.

Pinging @villebro for awareness given his current work on an improved security model.

Reply:
I do appreciate the discussion here, but this really feels like premature optimization to me.  The changes in this PR only affect two endpoints, and I don't predict that this kind of logic will needed in a widespread manner.

If I'm wrong and there is a larger need/desire for this kind of security measure, I definitely agree we should re-assess, but I personally don't see a need to come up with a more "perfect" solution at this point.

Reply:
> Ideally, you won't query for what you don't have access to, eliminating this concern because the values won't be in the session.

I'm not sure this is the right direction, imagine a situation where the backend needs to read a column for some reason, even though we don't want to surface that field to the logged-in user.  It seems wrong to tie the server's hands behind its back and not allow it to access the data just because of the current logged-in user.  I'd argue the API layer is where that kind of security check belongs, since it's sort of the gateway from the backend to the user.

Reply:
Here's a slight shift left solution, that's not quite to the DAO layer. Currently the DAO fetches the entire row, and then we're using Marshmallow to decide what is sent back to the client. Can we let Marshmallow decide what should/shouldn't be returned in the response schema since that is the mechanism that currently handles this responsibility?

Reply:
Yes. Ideally Marshmallow should be used to define both the request and response payload.

Additionally if this is going to reside in the API layer (at least in the interim) could we at least factor the code to adhere to the DRY principle?

Reply:
> I do appreciate the discussion here, but this really feels like premature optimization to me. The changes in this PR only affect two endpoints, and I don't predict that this kind of logic will needed in a widespread manner.

I thought that by referencing the concept of [Column Level Security (CLS)](https://blog.fabric.microsoft.com/en-us/blog/announcing-column-level-row-level-security-for-fabric-warehouse-sql-endpoint/) it would be more clear why this is not a premature optimization and why we will have the same need in a widespread manner.

> I'd argue the API layer is where that kind of security check belongs

If you research a little bit about CLS you'll see that many of these security checks are implemented at the database layer, even further from the API layer.

Reply:
I agree it's risky to implement this fix in a single interface, as it gives no guarantees that this information won't be exposed via some other interface, either pre-existing, or one that's added later. Mind you, this is not the only security concern regarding guest tokens - the issue of exposing granular information about the underlying query has been raised multiple times in the community, and is also something I feel we need to tackle soon.

Could we setup a meeting to discuss viable long-term solutions to these issues? I propose meeting to agree on the the following:
- A long term vision for fixing this and other similar security concerns. I agree with @michael-s-molina that baking this into the security work I've started sounds like a good idea. I can share what I have so far and my thoughts of how this can be taken forward.
- Agreeing on a quick fix to unblock heavy users of guest tokens to bridge the gap between now and when we have a proper security mechanism in place for handling this and other similar security issues.

Reply:
> I thought that by referencing the concept of [Object Level Security (OLS)](https://learn.microsoft.com/en-us/power-bi/enterprise/service-admin-ols) it would be more clear why this is not a premature optimization and why we will have the same need in a widespread manner.

This is referring to a security feature similar to row-level security where Superset could apply security filters to queries against connected databases, while the security issue at hand in this PR has to do with data from the Superset metadata DB.

> If you research a little bit about CLS you'll see that many of these security checks are implemented at the database layer, even further from the API layer.

My understanding is that CLS is for a security feature commonly implemented in data warehouses/BI tools, not a guide for how a server should implement security checks for data it owns

Reply:
I think it's a great idea to meet on long term solutions. If we were to move the logic for this endpoint into the schema (I left one more suggestion that would remove this from the api altogether) should we move forward with this for now as @villebro suggested?

Reply:
> I think it's a great idea to meet on long term solutions. If we were to move the logic for this endpoint into the schema (I left one more suggestion that would remove this from the api altogether) should we move forward with this for now as @villebro suggested?

@eschutho I also think moving this into the Marshmallow schema is probably the best solution for now.